Restaurant Reservation

raghad-ns opened the pull request.

raghad-ns committed:
Data models creation

raghad-ns committed:
fix data models naming

raghad-ns committed:
Database migration

raghad-ns committed:
Seeding database

raghad-ns committed:
Initializing entities repository with add, delete, and update methods

raghad-ns committed:
Add methods to query the DB
ListManagers(), GetReservationsByCustomer(CustomerId), ListOrdersAndMenuItems(ReservationId), ListOrderedMenuItems(ReservationId), CalculateAverageOrderAmount(EmployeeId)

raghad-ns committed:
Creating DB Function to calculate the total revenue generated by a specific restaurant

raghad-ns committed:
Adding reservation details view

raghad-ns committed:
Add employee restaurant details view
Creating data model for the view and query it

raghad-ns committed:
Creating stored procedure
FindCustomersHaveReservationWithPartySizeGreaterThan and create method in cutomer repository to execute it

raghad-ns committed:
Load app configurations from appsettings.json

WaleedJubeh commented on RestaurantReservation.Db/RestaurantReservationDbContext.cs:
```diff
@@ -1,6 +1,57 @@
+            entity.HasNoKey();
+        });
+
+        modelBuilder
```
Usually we create a migration script to seed the database

WaleedJubeh commented on RestaurantReservation.Db/RestaurantReservationDbContext.cs:
```diff
@@ -1,6 +1,57 @@
+    {
+        _connectionString = connString;
+    }
+    public DbSet<Customer> Customer { get; set; }
```
add a new line

WaleedJubeh commented:
Please apply the following Grouping configuration for the models

https://learn.microsoft.com/en-us/ef/core/modeling/#grouping-configuration

WaleedJubeh commented:
Please implmenet an interface for every repository and lets have them in RestaurantReservation project and lets move the repositories to the DB project

WaleedJubeh commented:
For Update/Add lets use the synchronous function like `_db.Customer.Add(customer);` and lets awat `SaveChangesAsync`

WaleedJubeh commented on RestaurantReservation/Customer/CustomerRepository.cs:
```diff
@@ -0,0 +1,40 @@
+
+    public async Task<Db.Models.Customer> UpdateCustomer(int customerId,Db.Models.Customer newCustomer)
+    {
+        var customer = await _db.Customer.FindAsync(customerId);
```
what if the item is not existed?
Please return null

WaleedJubeh commented on RestaurantReservation/Customer/CustomerRepository.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        var customer = await _db.Customer.FindAsync(customerId);
+        customer = newCustomer;
+        _db.SaveChangesAsync();
```
Lets use `Update`

WaleedJubeh commented on RestaurantReservation/Customer/CustomerRepository.cs:
```diff
@@ -0,0 +1,40 @@
+
+    public Task<List<Db.Models.Customer>> GetCustomersHaveReservationWithPartySizeGreaterThan(int partySize)
+    {
+        return _db.Customer.FromSqlInterpolated($"EXEC FindCustomersHaveReservationWithPartySizeGreaterThan {partySize}").ToListAsync();
```
Lets split it into 2 lines
```
    public Task<List<Db.Models.Customer>> GetCustomersHaveReservationWithPartySizeGreaterThan(int partySize)
    {
        return _db.Customer
            .FromSqlInterpolated($"EXEC FindCustomersHaveReservationWithPartySizeGreaterThan {partySize}")
            .ToListAsync();
    }
    ```

WaleedJubeh commented on RestaurantReservation.Db/Models/Employee.cs:
```diff
@@ -0,0 +1,14 @@
+    public int RestaurantId { get; set; } // Foreign key
+    public string FirstName { get; set; }
+    public string LastName { get; set; }
+    public string Position { get; set; }
```
Lets have enum

WaleedJubeh commented on RestaurantReservation.Db/RestaurantReservationDbContext.cs:
```diff
@@ -1,6 +1,57 @@
+    public DbSet<OrderItem> OrderItem { get; set; }
+    public DbSet<Reservation> Reservation { get; set; }
+    public DbSet<Restaurant> Restaurant { get; set; }
+    public DbSet<Table> Table { get; set; }
```
Lets have a DbSet for the view

WaleedJubeh commented on RestaurantReservation/Employee/EmployeeRepository.cs:
```diff
@@ -0,0 +1,46 @@
+
+    public Task<List<Db.Models.Employee>> ListManagers()
+    {
+        return _db.Employee.Where(emp => emp.Position.Equals("Manager")).ToListAsync();
```
Lets use the num here

WaleedJubeh commented on RestaurantReservation/Order/OrderRepository.cs:
```diff
@@ -0,0 +1,44 @@
+    {
+        return _db.Order
+            .Include(order => order.Items)
+            .ThenInclude(orderItem => orderItem.Item)
```
lets add another tab indentation

WaleedJubeh commented on RestaurantReservation/OrderItem/OrderItemRepository.cs:
```diff
@@ -0,0 +1,60 @@
+
+    public async Task<List<Db.Models.MenuItem>> ListOrderedMenuItemsForReservation(int reservationId)
+    {
+        var orders = await _db.Order
```
Can we simplify this query?

raghad-ns committed:
Getting rid of DB seeding logic from DB context class

raghad-ns committed:
Update repositories
Replacing AddAsync with Add, using .Update() for updates, and enhancing some queries

raghad-ns committed:
Adding repositories interfaces

raghad-ns committed:
Defining relationships between entities explicitly

raghad-ns committed:
Reverting previous changes
Moving relationships configuration to DBContext class

raghad-ns committed:
Moving repositories to DB project

raghad-ns committed:
Updating folder structure & adding Employee position enum

raghad-ns committed:
Moving application to web app
